### PR TITLE
Add GCM register module

### PR DESCRIPTION
**Pull request by tkchin**

### CODE REVIEW

**Comment by tkchin** [comment]
@jiayliu @chuckhays @juberti 
PTAL (taking ownership of CL)

File: samples/web/content/apprtc/gcm_register.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+PARAM_USER_ID_LIST = 'userIdList'
+
+class BindPage(webapp2.RequestHandler):
+  def get_message_from_json(self):
+    try:
+      message = json.loads(self.request.body)
+      return message
+    except Exception as e:
+      logging.warning('JSON load error from BindPage: error=' + str(e)
+          + ', request=' + self.request.body)
```
Comment (chuckhays) [inline samples/web/content/apprtc/gcm_register.py]: Not something we need to do in this change, but we should probably log more info about the failing request, like the URI, and anything else we know about the request. We probably need a "logging/diagnostics" feature to centralize the logging and provide a uniform log format.

File: samples/web/content/apprtc/gcm_register.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+      return None
+
+  def handle_new(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] and msg[PARAM_GCM_ID]:
```
Comment (chuckhays) [inline samples/web/content/apprtc/gcm_register.py]: Should this (and the other lines like it) use .has_key? I'm new to python but this appears to cause an error if the object doesn't have the key. Maybe we should centralize the message parsing/parameter retrieving into a helper.

File: samples/web/content/apprtc/gcm_register.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  def has(user_id, gcm_id):
+    q = GCMRecord.query(ancestor=get_ancestor_key()) \
+        .filter(GCMRecord.user_id == user_id).filter(GCMRecord.gcm_id == gcm_id)
+    return len(q.fetch()) > 0
+
+  @staticmethod
+  @ndb.transactional(retries=100)
+  def add_or_update(user_id, gcm_id, code):
+    q = GCMRecord.query(ancestor=get_ancestor_key()) \
+        .filter(GCMRecord.user_id == user_id).filter(GCMRecord.gcm_id == gcm_id)
```
Comment (chuckhays) [inline samples/web/content/apprtc/gcm_register.py]: It looks like we could put this query and the one on 128/129 into a helper.

File: samples/web/content/apprtc/gcm_register.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+      else:
+        self.response.out.write(constants.RESPONSE_UNVERIFIED)
+    else:
+      self.response.out.write(constants.RESPONSE_INVALID_ARGUMENT)
+
+  def handle_del(self):
+    msg = self.get_message_from_json()
+    if msg and msg[PARAM_USER_ID] and msg[PARAM_GCM_ID]:
+      GCMRecord.remove(msg[PARAM_USER_ID], msg[PARAM_GCM_ID])
+      self.response.out.write(constants.RESPONSE_SUCCESS)
```
Comment (chuckhays) [inline samples/web/content/apprtc/gcm_register.py]: Do we want to return SUCCESS in the case where we didn't find a matching record?

File: samples/web/content/apprtc/gcm_register.py (hunk 2 above)
Comment (tkchin) [inline samples/web/content/apprtc/gcm_register.py]: has_key is deprecated. Usual use is "x in <iter>". Moved to helper.

**Comment by tkchin** [comment]
PTAL

**Comment by tkchin** [comment]
PTAL

**Comment by jiayliu** [comment]
lgtm! Thx!

File: samples/web/content/apprtc/constants.py [hunk 5]
```diff
@@ -16,9 +16,19 @@
 
 WSS_HOST_PORT_PAIR = 'apprtc-ws.webrtc.org:443'
 
-RESPONSE_ERROR = 'ERROR'
-RESPONSE_ROOM_FULL = 'FULL'
-RESPONSE_UNKNOWN_ROOM = 'UNKNOWN_ROOM'
-RESPONSE_UNKNOWN_CLIENT = 'UNKNOWN_CLIENT'
-RESPONSE_DUPLICATE_CLIENT = 'DUPLICATE_CLIENT'
+RESPONSE_ERROR_INTERNAL = 'ERR_INTERNAL'
```
Comment (juberti) [inline samples/web/content/apprtc/constants.py]: RESPONSE_INTERNAL_ERROR sounds more natural

File: samples/web/content/apprtc/gcm_register.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+  return ndb.Key('GCMRecord', 'global')
+
+PARAM_USER_ID = 'userId'
+PARAM_GCM_ID = 'gcmId'
+PARAM_CODE = 'code'
+PARAM_OLD_GCM_ID = 'oldGcmId'
+PARAM_NEW_GCM_ID = 'newGcmId'
+PARAM_USER_ID_LIST = 'userIdList'
+
+def has_msg_field(msg, field):
```
Comment (juberti) [inline samples/web/content/apprtc/gcm_register.py]: Some function comments would be nice here.

File: samples/web/content/apprtc/gcm_register.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+        return message
+      logging.warning('Expected dictionary message'
+          + ', request=' + self.request.body)
+      return None
+    except Exception as e:
+      logging.warning('JSON load error from BindPage: error=' + str(e)
+          + ', request=' + self.request.body)
+      return None
+
+  def handle_new(self):
```
Comment (juberti) [inline samples/web/content/apprtc/gcm_register.py]: For future readability,

File: samples/web/content/apprtc/gcm_register.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+      if isinstance(message, dict):
+        return message
+      logging.warning('Expected dictionary message'
+          + ', request=' + self.request.body)
+      return None
+    except Exception as e:
+      logging.warning('JSON load error from BindPage: error=' + str(e)
+          + ', request=' + self.request.body)
+      return None
+
```
Comment (juberti) [inline samples/web/content/apprtc/gcm_register.py]: Function comments of the form
"Adds a new registration for a user ID and GCM ID, with a newly generated verification code."

File: samples/web/content/apprtc/gcm_register.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+                         code_sent_time = now,
+                         last_modified_time = now)
+      record.put()
+      logging.info(
+          'GCM binding added, user_id=%s, gcm_id=%s' % (user_id, gcm_id))
+      return constants.RESPONSE_CODE_SENT
+
+    assert len(records) == 1
+    if records[0].verified:
+      logging.warning('Cannot update GCM binding code since already verified, '\
```
Comment (juberti) [inline samples/web/content/apprtc/gcm_register.py]: Won't we still need to update the GCM ID post-verification? (e.g. on app upgrade)

Comment (juberti) [inline samples/web/content/apprtc/gcm_register.py]: Ah, I see that on the update path this is handled OK. NVm.

**Comment by juberti** [comment]
lgtm, please add some internal documentation as mentioned above

**Comment by tkchin** [comment]
PTAL

Comments added, also stricter type checking since we're already doing type checking.

**Comment by juberti** [comment]
lgtm

**Pull request merged**